ci-automation/garbage_collect_releases.sh

```diff
@@ -0,0 +1,180 @@
+
+                if (chan_num <= 2) {
+                    if (chan_count[chan_num] < keep)
+                        print version
+                    chan_count[chan_num] = chan_count[chan_num] + 1
+                } else {
+                    if (    (chan_count["lts"][major] < keep) \
+                         && (length(chan_count["lts"]) <= lts) )
+                        print version
+                    chan_count["lts"][major] = chan_count["lts"][major] + 1
+                }
+            } ')
+
+    echo
+    echo "######## The following version(s) will be kept ########"
+    if [ "$dry_run" = "y" ] ; then
+        echo
+        echo "(NOTE this is just a dry run since DRY_RUN=y)"
+        echo
+    fi
+    printf "%s\n" "${keep_versions[@]}"
+
+    local dir=""
+    for dir in  "sdk/amd64" \
+                "containers" \
+                "boards/amd64-usr" \
+                "boards/arm64-usr" \
+                "images/amd64" \
+                "images/arm64" \
+                "testing" \
+                ; do
+
+        local fullpath="${BUILDCACHE_PATH_PREFIX}/${dir}"
+        echo
+        echo "## Processing '${fullpath}'"
+        echo "---------------------------"
+        for version in $($sshcmd "${BUILDCACHE_USER}@${BUILDCACHE_SERVER}" \
+                                 "ls -1 ${BUILDCACHE_PATH_PREFIX}/${dir} | grep -E '^[0-9]+\.[0-9]+\.[0-9]+$'"); do
+            local o_fullpath="${fullpath}/${version}"
+
+            # skip if version is marked for keeping OR if it's a new release about to be published
+            if printf "%s\n" "${keep_versions[@]}" \
+               | { unset POSIXLY_CORRECT ; awk -v version="${version}" -v path="${dir}" '
+                BEGIN {
+                    vmajor = gensub("([0-9]+)\\.[0-9]+\\.[0-9]+","\\1","g", version) + 0
+                    vminor = gensub("[0-9]+\\.([0-9]+)\\.[0-9]+","\\1","g", version) + 0
+                    vpatch = gensub("[0-9]+\\.[0-9]+\\.([0-9]+)","\\1","g", version) + 0
+                    ret = 1
+                }
+
+                {
+                    if ($0 == version) {
+                        print ""
+                        print "## Skipping " version " because it is in the keep list."
+                        ret = 0
+                        exit
+                    }
+
+                    major = gensub("([0-9]+)\\.[0-9]+\\.[0-9]+","\\1","g") + 0
+                    minor = gensub("[0-9]+\\.([0-9]+)\\.[0-9]+","\\1","g") + 0
+                    patch = gensub("[0-9]+\\.[0-9]+\\.([0-9]+)","\\1","g") + 0
+
+                    if (   ((path == "sdk/amd64") || (path == "containers")) \
+                        && (vmajor == major) && (vminor == 0) && (vpatch == 0) ) {
+                        print ""
+                        print "## Skipping " version " in " path " because it contains the SDK for release " $0 " in keep list."
+                        ret = 0
+                        exit
+                    }
```

Comment:
That's tricky. We seldom had a case where we had to rebuild SDK for some patch release, so SDK ended up having a version different than `<MAJOR>.0.0`.

Reply:
Yeah, I agree, the check is too rigorous. We should skip all versions that have a path below `/sdk/smd64` and that match a major version on the keep list.